src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1312 @@
+            }
+        }
+
+        return this.Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap);
+    }
+
+    private ImmutableDictionary<ProjectId, ICompilationTracker> CreateCompilationTrackerMap(ProjectId changedProjectId, ProjectDependencyGraph dependencyGraph)
+    {
+        if (_projectIdToTrackerMap.Count == 0)
+            return _projectIdToTrackerMap;
+
+        using var _ = ArrayBuilder<KeyValuePair<ProjectId, ICompilationTracker>>.GetInstance(_projectIdToTrackerMap.Count, out var newTrackerInfo);
+        var allReused = true;
+        foreach (var (id, tracker) in _projectIdToTrackerMap)
+        {
+            var localTracker = tracker;
+            if (!CanReuse(id))
+            {
+                localTracker = tracker.Fork(tracker.ProjectState, translate: null);
+                allReused = false;
+            }
+
+            newTrackerInfo.Add(new KeyValuePair<ProjectId, ICompilationTracker>(id, localTracker));
+        }
+
+        if (allReused)
+            return _projectIdToTrackerMap;
+
+        return ImmutableDictionary.CreateRange(newTrackerInfo);
+
+        // Returns true if 'tracker' can be reused for project 'id'
+        bool CanReuse(ProjectId id)
+        {
+            if (id == changedProjectId)
+            {
+                return true;
+            }
+
+            return !dependencyGraph.DoesProjectTransitivelyDependOnProject(id, changedProjectId);
+        }
+    }
+
+    /// <inheritdoc cref="SolutionState.AddProject(ProjectInfo)"/>
+    public SolutionCompilationState AddProject(ProjectInfo projectInfo)
+    {
+        var newSolutionState = this.Solution.AddProject(projectInfo);
+        var newTrackerMap = CreateCompilationTrackerMap(projectInfo.Id, newSolutionState.GetProjectDependencyGraph());
+
+        return Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap);
+    }
+
+    /// <inheritdoc cref="SolutionState.RemoveProject(ProjectId)"/>
+    public SolutionCompilationState RemoveProject(ProjectId projectId)
+    {
+        var newSolutionState = this.Solution.RemoveProject(projectId);
+        var newTrackerMap = CreateCompilationTrackerMap(projectId, newSolutionState.GetProjectDependencyGraph());
+
+        return this.Branch(
+            newSolutionState,
+            projectIdToTrackerMap: newTrackerMap.Remove(projectId));
+    }
+
+    /// <inheritdoc cref="SolutionState.WithProjectAssemblyName"/>
+    public SolutionCompilationState WithProjectAssemblyName(
+        ProjectId projectId, string assemblyName)
+    {
+        return ForkProject(
+            this.Solution.WithProjectAssemblyName(projectId, assemblyName),
+            static (stateChange, assemblyName) => new CompilationAndGeneratorDriverTranslationAction.ProjectAssemblyNameAction(assemblyName),
+            forkTracker: true,
+            arg: assemblyName);
```

Comment:
many fork methods are even simpelr. they just pass the new solution state, no translation action, and a request to fork the tracker most of the time.